contracts/contracts/RaidenMicroTransferChannels.sol

```diff
@@ -172,57 +172,59 @@ contract RaidenMicroTransferChannels {
         require(token.transferFrom(msg.sender, address(this), _added_deposit));
     }
 
-    /// @notice Function called when any of the parties wants to close the channel and settle.
-    /// Receiver needs a balance proof to immediately settle, sender triggers a challenge period.
+    /// @notice Function called by the sender, receiver or a delegate, with all the needed
+    /// signatures to close the channel and settle immediately.
     /// @param _receiver_address The address that receives tokens.
     /// @param _open_block_number The block number at which a channel between the
     /// sender and receiver was created.
     /// @param _balance The amount of tokens owed by the sender to the receiver.
     /// @param _balance_msg_sig The balance message signed by the sender.
-    function uncooperativeClose(
+    /// @param _closing_sig The receiver's signed balance message, containing the sender's address.
+    function cooperativeClose(
         address _receiver_address,
         uint32 _open_block_number,
         uint192 _balance,
-        bytes _balance_msg_sig)
+        bytes _balance_msg_sig,
+        bytes _closing_sig)
         external
     {
-        address sender = verifyBalanceProof(_receiver_address, _open_block_number, _balance, _balance_msg_sig);
+        // Derive sender address from signed balance proof
+        address sender = extractBalanceProofSignature(_receiver_address, _open_block_number, _balance, _balance_msg_sig);
 
-        if(msg.sender == _receiver_address) {
-            settleChannel(sender, _receiver_address, _open_block_number, _balance);
-        } else {
-            require(msg.sender == sender);
-            initChallengePeriod(_receiver_address, _open_block_number, _balance);
-        }
+        // Derive receiver address from closing signature
+        address receiver = extractClosingSignature(sender, _open_block_number, _balance, _closing_sig);
```

Comment:
Is it possible for A and B to open channels with each other in both directions? If so, the `_balance_msg_sig` of one channel seems to be reusable as `_closing_sig` of the other.

It's a bit artificial, because in practice there won't be two channels between two parties opened in the same block, but maybe still worth considering.

Reply:
The signatures now differ by the `message_id` value and `address receiver` vs. `address sender` in the message hash. So you know that one is a balance proof and one is a closing signature and cannot swap them.

Reply:
Oops, yeah, you're right. Sorry, then it's fine.